File: AMR/Model/robot.py
import base64

__id: int = 1
__message: str = "Starting up"

# Position
__x_pos: float = 1
__y_pos: float = 1
__rotation: float = 0

# Destination
__x_destination: float = 1
__y_destination: float = 1

# Relevant info
__qr_scan: str = ""
__log = ["Jetbot is starting up"]
__user_command: str = ""
__robot_command: str = ""
__image = None

# Direction the robot is facing
__direction = {
    'Up': False,
    'Down': False,
    'Left': False,
    'Right': True
}
# Update direction
def turn_left():
    if __direction['Up'] is True:
        __direction['Up'] = False
        __direction['Left'] = True
        
    elif __direction['Down'] is True:
        __direction['Down'] = False
        __direction['Right'] = True
        
    elif __direction['Left'] is True:
        __direction['Left'] = False
        __direction['Down'] = True
        
    else:
        __direction['Right'] = False
        __direction['Up'] = True
        
# Update direction
def turn_right():
    if __direction['Up'] is True:
        __direction['Up'] = False
        __direction['Right'] = True
        
    elif __direction['Down'] is True:
        __direction['Down'] = False
        __direction['Left'] = True
        
    elif __direction['Left'] is True:
        __direction['Left'] = False
        __direction['Up'] = True
        
    else:
        __direction['Right'] = False
        __direction['Down'] = True

def get_direction():
    if __direction['Up'] is True:
        return 'Up'
    elif __direction['Down'] is True:
        return 'Down'
    elif __direction['Left'] is True:
        return 'Left'
    else:
        return 'Right'

        
def get_id():
    return __id


def get_message():
    return __message


def get_x_pos() -> float:
    return __x_pos


def get_y_pos() -> float:
    return __y_pos


def get_rotation() -> float:
    return __rotation


def get_x_destination() -> float:
    return __x_destination


def get_y_destination() -> float:
    return __y_destination


def get_qr_scan():
    return __qr_scan


def get_log():
    return __log


def get_user_command():
    return __user_command


def get_robot_command():
    return __robot_command


def get_image():
    return __image


def set_message(message: str):
    global __message
    __message = message


def set_x_pos(x_pos: float):
    global __x_pos
    __x_pos = x_pos


def set_y_pos(y_pos: float):
    global __y_pos
    __y_pos = y_pos


def set_rotation(rotation: float):
    global __rotation
    __rotation = rotation


def set_x_destination(x_destination: float):
    global __x_destination
    __x_destination = x_destination


def set_y_destination(y_destination: float):
    global __y_destination
    __y_destination = y_destination


def set_qr_scan(qr_scan: str):
    global __qr_scan
    __qr_scan = qr_scan


def set_log(log: [str]):
    global __log
    __log = log


def set_user_command(command: str):
    global __user_command
    __user_command = command


def set_robot_command(command: str):
    global __robot_command
    __robot_command = command


def set_image(image):
    global __image
    __image = image


def add_to_log(log):
    global __log
    __log.append(log)


def to_json():
    #b64 image for sending
    global __image
    base64_image = base64.b64encode(__image).decode('utf-8') if __image is not None else None

    return {
        'id': __id,
        'status': "Active",
        'message': __message,
        'x_pos': __x_pos,
        'y_pos': __y_pos,
        'rotation': __rotation,
        'x_destination': __x_destination,
        'y_destination': __y_destination,
        'qrScan': __qr_scan,
        'loggerList': __log,
        # 'userSignal': None,
        'image': base64_image
    }

File: server/AMRCapstone/src/main/java/com/AMRCapstone/AMRCapstone/model/Robot.java
package com.AMRCapstone.AMRCapstone.model;

import java.util.ArrayList;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.ObjectMapper;

import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;

@NotNull
@JsonInclude(JsonInclude.Include.ALWAYS)
public class Robot {
    // robot's id
    private int id;

    // status of robot.. i.e. "Active" / "Inactive" / "Connection Lost"
    @NotEmpty
    private String status;

    // a general overview of what the robot is currently doing.. i.e. "heading to QR
    // code"
    @NotEmpty
    private String message;

    // robot's position and rotation
    private float x_pos;
    private float y_pos;
    private float rotation;

    // robot's destination
    private float x_destination;
    private float y_destination;

    // qr code that has just been scanned
    private String qrScan = null;

    // list of robot's actions
    private ArrayList<String> loggerList;

    @JsonInclude(JsonInclude.Include.NON_NULL)
    private String userSignal = null;

    // image or video from robot's camera
    // object type need to be changed
    @NotEmpty
    private Byte[] image = null;

    public Robot(int id, String status, String message, float x_pos, float y_pos, float rotation,
            float x_destination, float y_destination, String qrScan, ArrayList<String> logger, Byte[] image) {
        this.id = id;
        this.status = status;
        this.message = message;
        this.x_pos = x_pos;
        this.y_pos = y_pos;
        this.rotation = rotation;
        this.x_destination = x_destination;
        this.y_destination = y_destination;
        this.qrScan = qrScan;
        this.loggerList = logger;
        this.image = image;
    }

    public Robot(int id, String status, String message, float x_pos, float y_pos, float rotation) {
        this.id = id;
        this.status = status;
        this.message = message;
        this.x_pos = x_pos;
        this.y_pos = y_pos;
        this.rotation = rotation;
        this.x_destination = 0;
        this.y_destination = 0;
        this.qrScan = null;
        this.loggerList = new ArrayList<String>();
        this.image = null;
    }

    public Robot() {

    }

    @JsonProperty("id")
    public int getId() {
        return id;
    }

    @JsonProperty("status")
    public String getStatus() {
        return status;
    }

    @JsonProperty("message")
    public String getMessage() {
        return message;
    }

    @JsonProperty("x_pos")
    public float getX_pos() {
        return x_pos;
    }

    @JsonProperty("y_pos")
    public float getY_pos() {
        return y_pos;
    }

    @JsonProperty("rotation")
    public float getRotation() {
        return rotation;
    }

    @JsonProperty("x_destination")
    public float getX_destination() {
        return x_destination;
    }

    @JsonProperty("y_destination")
    public float getY_destination() {
        return y_destination;
    }

    @JsonProperty("qrScan")
    public String getQrScan() {
        return qrScan;
    }

    @JsonProperty("loggerList")
    public ArrayList<String> getLoggerList() {
        return loggerList;
    }

    @JsonProperty("image")
    public Byte[] getImage() {
        return image;
    }

    @JsonProperty("userSignal")
    public String getUserSignal() {
        return userSignal;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public void setMessage(String message) {
        this.message = message;
    }

    public void setX_pos(float x_pos) {
        this.x_pos = x_pos;
    }

    public void setY_pos(float y_pos) {
        this.y_pos = y_pos;
    }

    public void setRotation(float rotation) {
        this.rotation = rotation;
    }

    public void setX_destination(float x_destination) {
        this.x_destination = x_destination;
    }

    public void setY_destination(float y_destination) {
        this.y_destination = y_destination;
    }

    public void setQrScan(String qrScan) {
        this.qrScan = qrScan;
    }

    public void setLoggerList(ArrayList<String> loggerList) {
        this.loggerList = loggerList;
    }

    public void addtoLoggerList(String action) {
        if (action == null)
            return;
        this.loggerList.add(action);
    }

    public void addtoLoggerList(ArrayList<String> loggerList) {
        if (loggerList == null)
            return;
        this.loggerList.addAll(loggerList);
    }

    public void setImage(Byte[] image) {
        if (image == null)
            return;
        this.image = image;
    }

    public void setUserSignal(String signal) {
        this.userSignal = signal;
    }

    @Override
    public String toString() {
        try {
            ObjectMapper objectMapper = new ObjectMapper();
            return objectMapper.writeValueAsString(this);
        } catch (Exception e) {
            e.printStackTrace();
            return "{}"; // Handle serialization exception
        }
    }
}
